[Java] Add a UTF-8 decoder that uses Unsafe to directly decode a byte buffer.

anuraaga opened the pull request:
While debugging performance of `UnsafeDirectNioDecoder` vs `ArrayDecoder`, I found that reading strings is a hotspot - profiles specifically showed `isValidUtf8` as a hotspot, and I think this is because the small string copies made while decoding a byte buffer cause poor utilization of the CPU cache compared to operating entirely on a contiguous heap array. The effect of this is that it is usually better to first copy an entire `ByteBuffer` into a `byte[]` and use `ArrayDecoder` for real-world protos that will have some strings - `UnsafeDirectNioDecoder` could pretty much never beat `ArrayDecoder` if there is any reasonable amount of string content in the proto.

This change adds a custom decoding path that works directly on byte buffers without copying onto the heap. The decoding code path is modeled after the current isValidUtf8 with actual codepoint calculation inserted. Performance goes up a lot and allows `UnsafeDirectNioDecoder` to be viable, especially when the strings are ASCII which is very common. This is nice for Netty users (i.e., gRPC), which usually work with direct buffers.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

xfxyjwf commented on java/core/src/main/java/com/google/protobuf/Utf8.java:
```diff
@@ -1554,5 +1767,93 @@ private static int unsafeIncompleteStateFor(long address, final int byte1, int r
+  /**
+   * Adaptation of http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
+   *
+   * Copyright (c) 2008-2009 Bjoern Hoehrmann <bjoern@hoehrmann.de>
```
I'll consult our internal team about this, but I think we can't accept this code piece unless the author 'Bjoern Hoehrmann' has signed google CLA and explicitly acknowledge the inclusion of his work in protobuf.

anuraaga commented:
@xfxyjwf Thanks for the quick initial review - was hoping that including the license would be sufficient but realize now that's not the case. I went ahead and reimplemented the logic by looking at Wikipedia + current isValidUtf8 code, so there isn't any licensed code anymore.

Fun enough, this new stateless version is faster than the other one even on non-ASCII - it seems to be faster in all cases so I went ahead and wired up the `byte[]` input streams too. Hopefully this can provide a good speedup for proto3 users across the board. Replicating JDK replacement logic is very hairy, so I've left that out, but doing so would also speedup proto2 users which could be pretty significant for large-scale legacy deployments.

Also removed JMH.

anuraaga commented:
@xfxyjwf Friendly ping - curious if the new approach looks reasonable

xfxyjwf commented:
@pherl @nmittler Can you guys also help take a look at this Java UTF-8 decoding changes? It looks good to me but it's a large change that I don't feel comfortable just merging without someone else also reviewing.

nmittler commented on java/core/src/main/java/com/google/protobuf/Utf8.java:
```diff
@@ -610,6 +634,98 @@ private static int partialIsValidUtf8(final ByteBuffer buffer, int index, final
+          if (offset >= limit - 1) {
+            throw InvalidProtocolBufferException.invalidUtf8();
+          }
+          byte byte2 = buffer.get(offset++);
```
It might be better to avoid the store operations and pass `buffer.get(offset++)` direction into the method call.

same comment throughout.

nmittler commented:
@anuraaga WRT JMH, I just saw the comment from @xfxyjwf ... I guess you'll follow this up with another PR for benchmarks?

pherl commented on java/core/src/main/java/com/google/protobuf/CodedInputStream.java:
```diff
@@ -825,13 +825,9 @@ public String readString() throws IOException {
     public String readStringRequireUtf8() throws IOException {
       final int size = readRawVarint32();
       if (size > 0 && size <= (limit - pos)) {
-        // TODO(martinrb): We could save a pass by validating while decoding.
```
Can we guard this change by a static boolean? That way it would be easier to roll back if there are performance regressions.

anuraaga replied:
Added a constant that reads a system property. It means javac won't remove the branches but JVM will easily do so, but let me know if you'd prefer a compiler constant.

pherl commented:
OK to test

anuraaga commented:
Ping @xfxyjwf @nmittler @pherl for another look if you have a chance. Looking forward to the speedup :)

nmittler commented:
@anuraaga can you re-run your benchmark to verify that the recent changes don't impact perf?

anuraaga commented:
@nmittler Yup, been running it on pretty much every small change since addressing comments and there's been no performance change while extracting the methods. Looked at the assembly with jitwatch and confirmed that it all inlines nicely.

nmittler commented:
@anuraaga Thanks! The changes all LGTM ... great work!

@xfxyjwf can you confirm that there is no perf regression on Android?

anuraaga commented:
Ping @xfxyjwf - any chance to check on Android? :) To be honest, wasn't thinking of Android at all so not sure what might happen there, but I guess AoT compiling should inline everything similarly.

anuraaga commented:
Hey @xfxyjwf @nmittler @pherl - any updates on this? Let me know if it would be smoother to split the flag into two, one for bytebuffers, defaulted `true`, and one for arrays, defaulted `false`.

- For bytebuffers, the performance advantage is massive and makes `UnsafeDirectNioDecoder` practical to use, so it's very important - IMO, fixing the bytebuffer performance issue is a bug fix more than an improvement. It's also the use case I'm personally invested in and which resulted in this PR.

- For arrays, the performance advantage also seems significant, but admittedly the potential effect is larger as it is the standard code path when using protobuf. While I'd like to apply it so people can benefit (I guess Google servers using proto3 would benefit :D ), I can see the need to take more caution.

nmittler commented:
@xfxyjwf It would be great to get this in.  I think we just need to make sure this doesn't have a negative impact on Android.  Can you verify that quickly?

xfxyjwf commented:
I'll run our internal microbenchmarks against this PR today.

nmittler commented:
Awesome! Thanks, @xfxyjwf!

xfxyjwf commented:
Benchmark results from our internal micro-benchmarks are neutral (both android ones and server ones). See: http://cl//176177299

anuraaga commented:
@xfxjwf Thanks! Just wondering are those proto3 benchmarks are a mix of proto2 and proto3? This change only affects proto3 (except for the rare case where someone turns on utf8 validation in proto2) so would want to look only at those for a change one way or another.

anuraaga commented:
Hello - another checkin :) I wish I could look at the benchmark as I'm curious as well, but I guess restricting to proto3, and also separating out ByteBuffer, is important to lessen noise. Is there anything I can do to help make progress on this? As always, no problem limiting the change to just UnsafeDirectNioDecoder if it helps.

anuraaga commented:
Hey @xfxyjwf @nmittler @pherl - checking in again (forgot it's Thanksgiving in the US last week, sorry for pinging on the holidays).

Was hoping that this would make it to 3.5 and a bit sad it didn't, but understand it's a big change. Any way we can move forward though (and possibly get in to 3.6 :D)?

In case there's some lack of enthusiasm since the internal benchmarks are neutral, I'm almost certain it's because they're mostly proto2 which is unaffected by this change. Throw a proto3 web search or maps suggest proto into there and I'm sure you'd see a big difference ;) I wouldn't be surprised if internal users are reluctant to switch to proto3 due to string decoding performance based on the numbers I'm seeing. But as always, I'm perfectly happy to restrict the change to direct byte buffers, which desperately needs this change, if it's easier to make progress.

FWIW, here are the numbers for the benchmark I've been using, on Java8 and Java9. Java9 doesn't get the 0-copy string move optimization, but even then it seems to be on-par with JDK for ASCII and much faster elsewhere.

The most important comparisons are
- StringsMessageBenchmark.byteBuffer / StringsMessageBenchmark.byteBuffer_copyFirst / StringsMessageBenchmark.byteBuffer_old
- ByteBufferBenchmark.decodeDirectly / DirectByteBufferBenchmark.copyToHeapFirst 
- ArrayBenchmark.array_custom / ArrayBenchmark.array_jdk

Source: https://gist.github.com/anuraaga/f94119780438bded4209f453a8f85169

```
Java 8

Benchmark                                                         (byteBufferType)   (value)   Mode  Cnt         Score        Error  Units
Utf8DecodeBenchmark.AllTypesArrayBenchmark.array_custom                        N/A       N/A  thrpt   20   1035675.946 ±  14019.777  ops/s
Utf8DecodeBenchmark.AllTypesArrayBenchmark.array_jdk                           N/A       N/A  thrpt   20    517516.528 ±  10718.457  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A   ENGLISH  thrpt   20  11344041.452 ± 140782.185  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A  JAPANESE  thrpt   20   4454176.927 ±  56869.938  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A     MIXED  thrpt   20   3146504.578 ±  45848.406  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A    HEBREW  thrpt   20   5882934.372 ± 100397.262  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A   CHINESE  thrpt   20   4475807.173 ±  66273.370  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A   ENGLISH  thrpt   20   8839746.140 ± 139353.873  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A  JAPANESE  thrpt   20   2661279.910 ±  32935.258  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A     MIXED  thrpt   20   1857703.502 ±  26763.854  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A    HEBREW  thrpt   20   3347970.823 ±  49345.380  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A   CHINESE  thrpt   20   3022370.706 ± 113706.644  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT   ENGLISH  thrpt   20   9631185.803 ± 159909.352  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT  JAPANESE  thrpt   20   4435904.935 ±  53276.172  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT     MIXED  thrpt   20   3297931.635 ±  65039.965  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT    HEBREW  thrpt   20   5600922.434 ±  95066.874  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT   CHINESE  thrpt   20   5140227.259 ±  87286.966  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP   ENGLISH  thrpt   20  11238772.624 ± 165250.902  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP  JAPANESE  thrpt   20   4432397.379 ±  44467.814  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP     MIXED  thrpt   20   3143543.713 ±  80910.752  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP    HEBREW  thrpt   20   5930799.888 ±  80753.351  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP   CHINESE  thrpt   20   4374172.489 ±  64001.325  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A   ENGLISH  thrpt   20   6673743.550 ± 109527.715  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A  JAPANESE  thrpt   20   2394081.304 ±  30274.810  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A     MIXED  thrpt   20   1620800.857 ±  32531.092  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A    HEBREW  thrpt   20   3106074.752 ±  41766.242  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A   CHINESE  thrpt   20   2758085.489 ±  44238.833  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT   ENGLISH  thrpt   20    416999.717 ±  16860.979  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT  JAPANESE  thrpt   20    152614.148 ±   3191.066  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT     MIXED  thrpt   20    105724.100 ±   1597.169  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT    HEBREW  thrpt   20    211967.029 ±   4796.024  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT   CHINESE  thrpt   20    150352.506 ±   1498.043  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT   ENGLISH  thrpt   20    368173.971 ±  13488.251  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT  JAPANESE  thrpt   20    163459.826 ±   4069.326  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT     MIXED  thrpt   20    115647.488 ±   2749.487  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT    HEBREW  thrpt   20    204758.891 ±   5319.577  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT   CHINESE  thrpt   20    169024.764 ±   5210.278  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP   ENGLISH  thrpt   20    459181.956 ±   5178.505  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP  JAPANESE  thrpt   20    158686.655 ±   2280.379  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP     MIXED  thrpt   20    108419.500 ±   2143.423  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP    HEBREW  thrpt   20    223132.476 ±   3852.461  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP   CHINESE  thrpt   20    159317.360 ±   2051.436  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT   ENGLISH  thrpt   20    381548.205 ±  14642.620  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT  JAPANESE  thrpt   20    140823.909 ±   3801.774  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT     MIXED  thrpt   20    101037.087 ±   2867.023  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT    HEBREW  thrpt   20    199506.587 ±   6117.641  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT   CHINESE  thrpt   20    142748.718 ±   3096.335  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP   ENGLISH  thrpt   20    424750.212 ±   7060.122  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP  JAPANESE  thrpt   20    145176.969 ±  10440.393  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP     MIXED  thrpt   20    104181.653 ±   1632.933  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP    HEBREW  thrpt   20    207804.030 ±   8557.704  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP   CHINESE  thrpt   20    155530.760 ±   2465.193  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT   ENGLISH  thrpt   20    273278.066 ±  11323.380  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT  JAPANESE  thrpt   20     84605.972 ±   1433.077  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT     MIXED  thrpt   20     59695.944 ±    858.012  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT    HEBREW  thrpt   20    109256.426 ±   2135.230  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT   CHINESE  thrpt   20    101906.669 ±   3213.892  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP   ENGLISH  thrpt   20    377401.924 ±   6666.030  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP  JAPANESE  thrpt   20     95649.830 ±   1185.386  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP     MIXED  thrpt   20     65389.237 ±    919.510  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP    HEBREW  thrpt   20    125759.850 ±   2539.549  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP   CHINESE  thrpt   20    114718.268 ±   1671.434  ops/s

Java 9

Benchmark                                                         (byteBufferType)   (value)   Mode  Cnt         Score        Error  Units
Utf8DecodeBenchmark.AllTypesArrayBenchmark.array_custom                        N/A       N/A  thrpt   20    979233.661 ±  10012.953  ops/s
Utf8DecodeBenchmark.AllTypesArrayBenchmark.array_jdk                           N/A       N/A  thrpt   20    541470.347 ±   9336.800  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A   ENGLISH  thrpt   20   9996077.522 ± 125613.379  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A  JAPANESE  thrpt   20   4209367.121 ±  55791.397  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A     MIXED  thrpt   20   2883566.824 ±  50564.694  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A    HEBREW  thrpt   20   5685211.678 ±  67744.823  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_custom                                N/A   CHINESE  thrpt   20   4378542.460 ± 111577.429  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A   ENGLISH  thrpt   20  10687035.954 ± 157215.156  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A  JAPANESE  thrpt   20   2826209.172 ±  39353.465  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A     MIXED  thrpt   20   2025476.087 ±  21770.999  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A    HEBREW  thrpt   20   3477542.374 ±  44940.345  ops/s
Utf8DecodeBenchmark.ArrayBenchmark.array_jdk                                   N/A   CHINESE  thrpt   20   3325769.121 ±  42774.892  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT   ENGLISH  thrpt   20   8822377.021 ± 207093.640  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT  JAPANESE  thrpt   20   3997758.356 ± 452676.034  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT     MIXED  thrpt   20   3210459.933 ± 120734.598  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT    HEBREW  thrpt   20   5115177.256 ±  73427.771  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                      DIRECT   CHINESE  thrpt   20   4364487.868 ±  76792.000  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP   ENGLISH  thrpt   20  10105268.129 ± 124063.026  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP  JAPANESE  thrpt   20   4202349.863 ±  70464.257  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP     MIXED  thrpt   20   2992099.443 ±  52333.824  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP    HEBREW  thrpt   20   5603021.148 ± 105407.183  ops/s
Utf8DecodeBenchmark.ByteBufferBenchmark.decodeDirectly                        HEAP   CHINESE  thrpt   20   4315432.381 ±  63811.573  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A   ENGLISH  thrpt   20   8015844.856 ± 155062.216  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A  JAPANESE  thrpt   20   2440636.535 ±  32144.760  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A     MIXED  thrpt   20   1656699.029 ±  29489.713  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A    HEBREW  thrpt   20   2941237.104 ±  86575.047  ops/s
Utf8DecodeBenchmark.DirectByteBufferBenchmark.copyToHeapFirst                  N/A   CHINESE  thrpt   20   2840008.086 ±  46602.012  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT   ENGLISH  thrpt   20    380879.953 ±  18276.071  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT  JAPANESE  thrpt   20    142730.085 ±   1912.097  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT     MIXED  thrpt   20     99962.482 ±   1479.362  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT    HEBREW  thrpt   20    188939.494 ±   4516.471  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.array                           DIRECT   CHINESE  thrpt   20    159535.449 ±   1936.158  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT   ENGLISH  thrpt   20    357865.915 ±   8309.979  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT  JAPANESE  thrpt   20    154177.425 ±   2456.391  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT     MIXED  thrpt   20    112304.389 ±   3628.963  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT    HEBREW  thrpt   20    186148.594 ±   2943.913  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                      DIRECT   CHINESE  thrpt   20    162703.751 ±   3820.713  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP   ENGLISH  thrpt   20    399438.689 ±   3683.416  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP  JAPANESE  thrpt   20    150696.624 ±   4219.753  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP     MIXED  thrpt   20    102825.498 ±   1643.060  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP    HEBREW  thrpt   20    211749.172 ±   4625.593  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer                        HEAP   CHINESE  thrpt   20    161054.281 ±   2183.351  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT   ENGLISH  thrpt   20    318984.779 ±   8135.769  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT  JAPANESE  thrpt   20    134186.007 ±   4598.491  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT     MIXED  thrpt   20     95816.538 ±   1283.384  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT    HEBREW  thrpt   20    178311.424 ±   7073.692  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst            DIRECT   CHINESE  thrpt   20    138070.212 ±   2569.738  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP   ENGLISH  thrpt   20    375249.554 ±   7202.198  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP  JAPANESE  thrpt   20    142648.957 ±   2582.319  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP     MIXED  thrpt   20     97951.935 ±   1311.572  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP    HEBREW  thrpt   20    198253.487 ±   2770.646  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_copyFirst              HEAP   CHINESE  thrpt   20    152498.568 ±   2756.298  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT   ENGLISH  thrpt   20    367151.701 ±   6440.305  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT  JAPANESE  thrpt   20     88343.758 ±   2391.535  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT     MIXED  thrpt   20     64460.688 ±   1533.509  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT    HEBREW  thrpt   20    108406.855 ±   2833.226  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                  DIRECT   CHINESE  thrpt   20    102713.102 ±   1617.851  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP   ENGLISH  thrpt   20    440636.287 ±   2489.340  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP  JAPANESE  thrpt   20     93893.509 ±   1821.500  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP     MIXED  thrpt   20     67751.734 ±   1045.973  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP    HEBREW  thrpt   20    122597.875 ±   1997.821  ops/s
Utf8DecodeBenchmark.StringsMessageBenchmark.byteBuffer_old                    HEAP   CHINESE  thrpt   20    115595.428 ±   1846.056  ops/s
```

xfxyjwf commented:
@anuraaga Sorry for the late response. Yes, our internal benchmarks are proto2 heavy. The PR should be fine as long as it doesn't negatively impact proto2 performance. We will need to add more benchmarks for proto3 protos but it's not a blocking issue for this PR.

I'll wait for @nmittler @pherl in case they have some more comments and if not I'll merge it tomorrow.

anuraaga commented:
Thanks for the reviews!

On Dec 1, 2017 04:05, "Feng Xiao" <notifications@github.com> wrote:

> Merged #3824 <https://github.com/google/protobuf/pull/3824>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/protobuf/pull/3824#event-1366303763>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMGyGhhzHZ5QVOj75qYMvwrgjfkv0r8ks5s7vxbgaJpZM4QMRor>
> .
>

xfxyjwf commented on java/core/src/main/java/com/google/protobuf/UnsafeUtil.java:
```diff
@@ -259,6 +260,26 @@ static long addressOffset(ByteBuffer buffer) {
+      // This should never happen, but return a copy as a fallback just in case.
+      return new String(chars);
+    }
+    putObject(str, STRING_VALUE_OFFSET, chars);
```
Copy-paste the comment from Luke:

Seems like this undermines the memory semantics of the value field.

The value field in string is marked final, but this is a non-volatile write.  So if this string is unsafely published (which is possible using the APIs this is exposed through), then people could get NPEs when calling methods on this string.

There is also the other problem about the other fields in string.  In Java7/8 it doesn't have any (besides hashCode), but in the past it has had a length and offset field and in the future it might have other fields.  How do we make sure that this code works on future JDK versions, or JDKs with different field semantics.  Or things like -XX:+UseCompressedStrings in jdk6 which changed the field to be a byte[] instead of a char[]

For example in jdk10 it looks like there is an extra byte field
http://hg.openjdk.java.net/jdk10/master/file/be620a591379/src/java.base/share/classes/java/lang/String.java


To fix the memory ordering problem you can probably just use putObjectVolatile, as for handling compatibility with multiple future jdk versions, im not sure.

anuraaga replied:
I thought the memory ordering would work because we both allocate and populate the variable in this same function - it should not be possible to dereference `value` until the object is initialized (this code should be effectively the same as the code in the standard constructor). But perhaps inlining + escape analysis breaks that assumption - are you seeing that? Doing a volatile write would indeed fix that.

As for JDK9+, they don't use a `char[]` for `value` so don't execute this code (guarded by the first if). I figured this would keep it future proof for all practical purposes, and in the fairly inconveivable chance that `char[] value` is restored in a future version a change could be made then. The speedup on JRE8 is reasonable, but not a deal breaker so it would be ok to remove this function too.

lukesandberg replied:
The memory semantics of final fields and constructors are fairly well documented in JLS
https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4

the key thing is that final field writes force a memory barrier at the end of the constructor.  But since you aren't using the constructor this doesn't work.  I believe volatile will fix the problem in the current code in JDK8, but it can still cause other problems in the future. 

I think the function should be removed and the string constructor should simply be invoked.  The caller of this function actually performs an Arrays.copyOf to correctly size the array.  instead we could just call the String(char[], int, int) constructor and do both things at the same time.  At least in some cases it would be just as good as the current unsafe code

anuraaga replied:
Ah yeah, I was thinking of the function return here as a memory fence for some reason, but obviously that's not correct.

I still think the chance of a problem in future Java versions is extremely low since the method is guarded by the presence of `char[] value`, but at the same time that's enough reason to avoid this  hacky optimization if it only applies to older (Java8) runtimes. Will send a PR
